module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php

```diff
@@ -0,0 +1,289 @@
+     * @param string             $requestedName Service being created
+     * @param null|array         $options       Extra options (optional)
+     *
+     * @return object
+     *
+     * @throws ServiceNotFoundException if unable to resolve the service.
+     * @throws ServiceNotCreatedException if an exception is raised when
+     * creating a service.
+     * @throws ContainerException&\Throwable if any other error occurs
+     */
+    public function __invoke(
+        ContainerInterface $container,
+        $requestedName,
+        array $options = null
+    ) {
+        if (!empty($options)) {
+            throw new \Exception('Unexpected options sent to factory.');
+        }
+        $yamlReader = $container->get(\VuFind\Config\YamlReader::class);
+        $oauth2Config = $yamlReader->get('OAuth2Server.yaml');
+
+        $this->container = $container;
+        $this->oauth2Config = $oauth2Config;
+        $tablePluginManager = $this->container
+            ->get(\VuFind\Db\Table\PluginManager::class);
+        $this->accessTokenTable = $tablePluginManager->get('AccessToken');
+
+        $session = new \Laminas\Session\Container(
+            OAuth2Controller::SESSION_NAME,
+            $container->get(\Laminas\Session\SessionManager::class)
+        );
+        return $this->applyPermissions(
+            $container,
+            new $requestedName(
+                $container,
+                $oauth2Config,
+                $this->getAuthorizationServerFactory(),
+                $this->getResourceServerFactory(),
+                $container->get(\LmcRbacMvc\Service\AuthorizationService::class),
+                $container->get(\VuFind\Validator\CsrfInterface::class),
+                $session,
+                $this->getIdentityRepository(),
+                $this->accessTokenTable
+            )
+        );
+    }
+
+    /**
+     * Return a factory function for creating the authorization server.
+     *
+     * @return callable
+     */
+    protected function getAuthorizationServerFactory(): callable
+    {
+        return function (?string $clientId): AuthorizationServer {
+            if (!($keyPath = $this->oauth2Config['Server']['privateKeyPath'] ?? '')
+            ) {
+                throw new \Exception(
+                    'Server/privateKeyPath missing from OAuth2Server.yaml'
+                );
+            }
+            if (strncmp($keyPath, '/', 1) !== 0) {
+                // Convert relative path:
+                $keyPath = Locator::getConfigPath($keyPath);
+            }
+            $encryptionKey
+                = trim($this->oauth2Config['Server']['encryptionKey'] ?? '');
+            if (!$encryptionKey) {
+                throw new \Exception(
+                    'Server/encryptionKey missing from OAuth2Server.yaml'
+                );
+            }
+
+            $server = new AuthorizationServer(
+                new ClientRepository($this->oauth2Config),
```

Comment:
Should we have separate getters (or perhaps properties containing class names) for the various repositories and other VuFind-specific classes so it's easier to locally override them?

Reply:
I added some getters and then decided that it's better to actually utilize the service manager, so now it all goes through it. I also split some methods to smaller parts for easier overriding.

Reply:
Thanks, @EreMaijala, this is definitely cleaner now. I did momentarily wonder if we should have an Oauth2-specific plugin manager to avoid putting so many narrow-purpose definitions into the top-level service manager, but given that these classes don't have a common base class or interface, I think what you've done makes the most sense. If we wanted to separate these concerns from the rest of the code, maybe it would actually make the most sense to factor them to a separate VuFind\Oauth2 Laminas module... but I don't think that's something we need to contemplate right now, unless it appeals to you.

Reply:
I thought about a plugin manager, but it didn't feel appropriate. A separate module is worth a thought, but I feel it would be so intertwined with the base module that it gets complicated easily, particularly if the functionality is expanded in the future so that it'd be possible to remember and manage access grants.

Reply:
Agreed -- worth consideration, but not an urgent priority right now. We can see how this evolves and refactor later if justified.